python/rikai/spark/sql/codegen/mlflow_registry.py

```diff
@@ -0,0 +1,303 @@
+    registered_model_name: Optional[str] = None,
+    **kwargs,
+):
+    """Convenience function to log the model with information needed by rikai.
+    This should be called during training when the model artifact is produced.
+
+
+    Parameters
+    ----------
+    model: Any
+        The model artifact object
+    artifact_path: str
+        The relative (to the run) artifact path
+    flavor: str
+        A valid Mlflow flavor (e.g., "pytorch")
+    schema: str
+        Output schema (pyspark DataType)
+    pre_processing: str, default None
+        Full python module path of the pre-processing transforms
+    post_processing: str, default None
+        Full python module path of the post-processing transforms
+    registered_model_name: str, default None
+        Model name in the mlflow model registry
+    kwargs: dict
+        Passed to `mlflow.<flavor>.log_model`
+
+    Examples
+    --------
+    # import mlflow.pytorch  <= vanilla mlflow usage
+    from rikai.spark.sql.codegen.mlflow_registry import log_model
+
+    # Log PyTorch model
+    with mlflow.start_run() as run:
+
+        # Training loop
+        # ...
+
+        # Assume `model` is the trained model from the training loop
+        log_model(model, "model", "pytorch",
+                  schema=rikai.contrib.torch.transforms.fasterrcnn_resnet50_fpn.OUTPUT_SCHEMA,
+                  pre_processing="rikai.contrib.torch.transforms.fasterrcnn_resnet50_fpn.pre_processing",
+                  post_processing="rikai.contrib.torch.transforms.fasterrcnn_resnet50_fpn.post_processing",
+                  registered_model_name="MyPytorchModel")
+        # mlflow.pytorch.log_model(model, "model") <= vanilla mlflow usage
+
+
+    For more details see `mlflow docs <https://www.mlflow.org/docs/latest/python_api/mlflow.pytorch.html#mlflow.pytorch.log_model>`_.
+    """  # noqa E501
+    # no need to set the tracking uri here since this is intended to be called
+    # inside the training loop within mlflow.start_run
+    getattr(mlflow, flavor).log_model(
+        model,
+        artifact_path,
+        registered_model_name=registered_model_name,
+        **kwargs,
+    )
+    tags = {
+        "rikai.spec.version": "1.0",
+        "rikai.model.flavor": flavor,
+        "rikai.output.schema": schema,
+        "rikai.transforms.pre": pre_processing,
+        "rikai.transforms.post": post_processing,
+        "rikai.model.artifact_path": artifact_path,
+    }
+    mlflow.set_tags(tags)
+
+
+class MlflowRegistry(Registry):
+    """MLFlow-based Model Registry"""
+
+    def __init__(self, spark: SparkSession):
+        self._spark = spark
+        self._jvm = spark.sparkContext._jvm
+
+    def __repr__(self):
+        return "MlflowRegistry"
+
+    def resolve(self, uri: str, name: str, options: Dict[str, str]):
+        logger.info(f"Resolving model {name} from {uri}")
+        tracking_uri = self._spark.conf.get(CONF_MLFLOW_TRACKING_URI)
+        client = MlflowClient(tracking_uri)
```

Comment:
should we keep a `self._client`? `resolve` will be called everytime a `CREATE MODEL` is called.

Reply:
it's not heavy weight but sure I can do that. Probably need to check if the tracking uri changes and then re-instantiate the client